Notes/Proposals/Explicit-partial-type-inference.md

```diff
@@ -0,0 +1,372 @@
+```
+
+> [12.6.3.11 Upper-bound inferences](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#126311-upper-bound-inferences) - Propagting the constraints to type variables contained already in bounds
+
+```diff
+-If V is one of the unfixed Xᵢ then U is added to the set of exact bounds for Xᵢ.
++If P is one of the unfixed Pᵢ then U is added to the set of upper bounds for Pᵢ.
++ For each upper bound B_u of P_i containing at least one unfixed type variable P_k, an upper-bound inference is made from U to B_u
++ For each exact bound B_e of P_i containing at least one unfixed type variable P_k, an upper-bound inference is made from U to B_e
++ For each lower bound B_l of P_i containing at least one unfixed type variable P_k, an upper-bound inference is made from U to B_l
+```
+
+> [12.6.3.12 Fixing](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#126312-fixing) - Substituting fixed variables in bounds to fixed results.
+
+```diff
+- Rename X_i
++ to P_i refferring extended list of type variables by inferred type variables
+- The set of candidate types Uₑ starts out as the set of all types in the set of bounds for Xᵢ
++ The set of candidate types Uₑ starts out as the set of all types in the set of bounds for Vᵢ with substituted fixed type variables.
+```
+
+> [12.6.5 Compile-time checking of dynamic member invocation](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1265-compile-time-checking-of-dynamic-member-invocation) - Prohibiting inferred type arguments with dynamic
+
+```diff
++Because type inference when `dynamic` is used is done in run-time, we prohibit using inferred type argument in the type argument list, because the run-time doesn't support it.
+Even though overload resolution of a dynamically bound operation takes place at run-time, it is sometimes possible at compile-time to know the list of function members from which an overload will be chosen:
+```
+
+> Note: In the current situation, it is impossible that bounds of a type variable contain the type variable itself or even create a cycle. 
+> However, it can happen when we add other contraints like `where` clauses.
+>
+> How to solve it ?
+>
+>Probably, we can wait until there will be no type variables, which we can fix accordingly to the conditions described above. 
+>Then check whether there is a non-dependent type variable containing at least one bound which doesn't contain any unfixed type variables, hide the bounds which are inferred and do the fix.
+> Although, I'm not sure if it is a correct solution.
+
+#### Object construction
+
+> TODO
+
+#### Variable declaration
+
+> TODO
+
+#### Casting
+
+> TODO
+
+## Alternatives
+
+We can choose different char or keywords for determining compiler inferred type argument. 
+Other options consist of `*`, `var`, `?`, `<<whitespace>>`, or no token at all (e.g. `Foo<,>(..,)`). 
```

Comment:
One reason to pick a specific token is that it may be generalized to more places types are used. For example, `_[] local = new[] { 1, 2, 3 };`, while it doesn't look very nice, could work syntactically, but `[] local = ...` probably wouldn't, at least when we look across all the different places we'd like such types to appear.